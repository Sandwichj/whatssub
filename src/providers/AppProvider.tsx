import * as Font from 'expo-font';

import React, { useEffect, useReducer } from 'react';

import { AsyncStorage } from 'react-native';
import { ThemeType } from '../types';

interface Context {
  state: State;
  changeTheme: (theme?: ThemeType) => void;
}
const AppContext = React.createContext<Context>(null);
const AppConsumer = AppContext.Consumer;

interface ChangeThemeTypeAction {
  type: 'change-theme-type-mode';
  payload: {
    theme?: ThemeType;
  };
}

interface Props {
  theme?: ThemeType;
  children?: any;
}

interface State {
  theme?: ThemeType;
}

export const initialState: State = {
  theme: ThemeType.LIGHT,
};

const reducer = (state: State, action: ChangeThemeTypeAction) => {
  switch (action.type) {
    case 'change-theme-type-mode':
      // if payload.theme doesn't exist, toggle
      if (!action.payload.theme) {
        return {
          ...state,
          theme:
            state.theme === ThemeType.LIGHT
              ? ThemeType.DARK
              : ThemeType.LIGHT,
        };
      }
      // else change it with payload.theme
      return {
        ...state,
        theme: action.payload.theme,
      };
    default:
      return state;
  }
};

function AppProvider(props: Props) {
  const [state, dispatch] = useReducer(reducer, initialState);

  // by encapsulating dispatch, it is easier to use (such as passing through screenProps)
  const changeTheme: (theme?: ThemeType) => void = (theme) => {
    dispatch({
      type: 'change-theme-type-mode',
      payload: {
        theme,
      },
    });
  };

  // initialize theme
  useEffect(() => {
    changeTheme(props.theme ? props.theme : initialState.theme);
  }, [props.theme]);

  // Load font and then use saved theme from local storage
  const getCurrentThemeType = async () => {
    const value = (await AsyncStorage.getItem('theme')) as ThemeType;
    value && changeTheme(value);
  };
  useEffect(() => {
    Font.loadAsync({
      'spoqa-han-sans-bold': require('../../assets/fonts/SpocaHanSans/SpoqaHanSans-Bold.ttf'),
      'spoqa-han-sans-regular': require('../../assets/fonts/SpocaHanSans/SpoqaHanSans-Regular.ttf'),
    });
    getCurrentThemeType();
  }, []);

  useEffect(() => {
    AsyncStorage.setItem('theme', state.theme);
  }, [state.theme]);

  return (
    <AppContext.Provider
      value={{
        state,
        changeTheme,
      }}
    >
      {props.children}
    </AppContext.Provider>
  );
}

export { AppConsumer, AppProvider, AppContext };
