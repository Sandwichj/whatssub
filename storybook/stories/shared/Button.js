import React from 'react';
import { storiesOf, addDecorator } from '@storybook/react-native';
import Button from '../../../src/components/shared/Button';
import styled from 'styled-components/native';
import { IC_LOGO, IC_GOOGLE, IC_FACEBOOK, IC_SLASH } from '../../../src/utils/Icons';
import { text, boolean, number } from '@storybook/addon-knobs';

const Container = styled.View`
  flex: 1;
  justify-content: center;
  background-color: ${({ theme }) => theme.backgroundLight};
`;

const ContainerDeco = (storyFn) => (
  <Container>
    {storyFn()}
  </Container>
);

storiesOf('shared-Button', module)
  .addDecorator(ContainerDeco)
  .add('default', () => (
    <Button
      onClick={() => console.log('button clicked')}
      text='Test Button'
      style={{
        backgroundColor: '#fff',
      }}
    />
  ))
  .add('isLoading', () => (
    <Button
      isLoading={true}
    />
  ))
  .add('isDisabled', () => (
    <Button
      text='Test Button'
      isDisabled={true}
    />
  ))
  .add('img Source', () => (
    <>
      <Button
        text='Test Button'
        style={{
          backgroundColor: '#fff',
        }}
        imgLeftSrc={IC_GOOGLE}
      />
      <Button
        text='Test Button'
        style={{
          backgroundColor: '#fff',
        }}
        imgLeftSrc={IC_FACEBOOK}
      />
    </>
  ))
  .add('custom', () => (
    <Button
      style={{
        backgroundColor: '#ccc',
      }}
      text={text('text', 'editable')}
      isDisabled={boolean('isDisabled', false)}
      isLoading={boolean('isLoading', false)}
    />
  ));
