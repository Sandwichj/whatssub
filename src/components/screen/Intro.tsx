import React from 'react';
import firebase from 'firebase';
import { useAuthState } from 'react-firebase-hooks/auth';

import {
  View,
} from 'react-native';
import { Text } from 'react-native-animatable';
import { NavigationScreenProp, NavigationStateRoute } from 'react-navigation';
import { AppProvider as Provider, AppConsumer, AppContext } from '../../providers';

import styled from 'styled-components/native';

import { ThemeType } from '../../theme';
import { IC_LOGO, IC_GOOGLE, IC_FACEBOOK, IC_SLASH } from '../../utils/Icons';
import { getString } from '../../../STRINGS';
import Button from '../shared/Button';
import useInterval from '../../hooks/useInterval';

const Container = styled.View`
  flex: 1;
  align-self: stretch;
  overflow: scroll;
  background-color: ${({ theme }) => theme.backgroundDark};

  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  overflow: hidden;
`;

const TitleWrapper = styled.View`
  margin-top: 120;
  align-self: center;

  flex-direction: row;
  justify-content: center;
  align-items: center;
  text-align: center;
`;

const LogoImage = styled.Image`
  width: 135;
  height: 30;
`;

const SlashImage = styled.Image`
  width: 14;
  height: 14;
`;

const ContentWrapper = styled.View`
  flex-direction: column;
  height: 100%;
  width: 100%;
  justify-content: center;
  align-items: flex-start;
`;

const ContentScroll = styled.ScrollView`
  height: 100%;
  width: 100%;
`;

const StyledAnimatableText = styled(Text)`
  font-size: 36;
  font-weight: 600;
  min-height: 44;
  line-height: 48;
  color: white;
`;

const StyledText = styled.Text`
  font-size: 36;
  font-weight: 600;
  line-height: 48;
  text-align: center;
  color: ${({ theme }) => theme.fontLight};
`;

const ButtonWrapper = styled.View`
  bottom: 40;
  width: 100%;
  height: 200;
  background-color: white;
  margin-top: 28;
  padding-top: 28;

  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
`;

interface IProps {
  store?: any;
  screenProps?: any;
  navigation?: NavigationScreenProp<NavigationStateRoute<any>>;
}

function Intro(props: IProps) {
  const titleArray = [
    getString('INTRO_TITLE_1'),
    getString('INTRO_TITLE_2'),
    getString('INTRO_TITLE_3'),
    getString('INTRO_TITLE_4'),
    getString('INTRO_TITLE_5'),
  ];
  const [titleIndex, setTitleIndex] = React.useState(0);
  // const [user, initialising, error] = useAuthState(firebase.auth());

  // const changeTheme = () => {
  //   let payload: object;
  //   if (state.theme === ThemeType.LIGHT) {
  //     payload = {
  //       theme: ThemeType.DARK,
  //     };
  //   } else {
  //     payload = {
  //       theme: ThemeType.LIGHT,
  //     };
  //   }
  //   dispatch({
  //     type: 'change-theme-mode',
  //     payload,
  //   });
  // };

  const googleLogin = () => {
    console.log('googleLogin');
  };

  const facebookLogin = () => {
    console.log('facebookLogin');
  };

  const btnStyle = {
    backgroundColor: 'white',
    borderWidth: 1,
    borderRadius: 0,
    borderColor: props.screenProps.theme.btnWhiteBorder,
  };

  useInterval(() => {
    if (titleIndex === 4) {
      setTitleIndex(0);
    } else {
      setTitleIndex(titleIndex + 1);
    }
  }, 1000);

  return (
    <Container>
      <ContentWrapper>
        <ContentScroll
          contentContainerStyle={{
            justifyContent: 'center',
            alignItems: 'center',
          }}
        >
          <TitleWrapper>
            <LogoImage source={IC_LOGO}/>
          </TitleWrapper>
          <SlashImage
            style={{
              marginTop: 68,
              marginBottom: 34,
            }}
            source={IC_SLASH}
          />
          <StyledAnimatableText
            animation='fadeIn'
            iterationCount='infinite'
            direction='alternate'
            duration={1000}
          >
            { titleArray[titleIndex] }
          </StyledAnimatableText>
          <StyledText>{getString('INTRO_MESSAGE')}</StyledText>
          <StyledText style={{
            color: props.screenProps.theme.fontTint,
          }}>{getString('INTRO_WHATSSUB')}</StyledText>
          <SlashImage
            style={{
              marginTop: 38,
              marginBottom: 68,
            }}
            source={IC_SLASH}
          />
          <ButtonWrapper>
            <Button
              testID='btnGoogle'
              style={btnStyle}
              imgLeftSrc={IC_GOOGLE}
              onClick={ () => googleLogin() }
              text={getString('SIGN_IN_WITH_GOOGLE')}
            />
            <View style={{ marginTop: 8 }}/>
            <Button
              testID='btnFacebook'
              style={btnStyle}
              imgLeftSrc={IC_FACEBOOK}
              imgLeftStyle={{
                height: 28,
                width: 16,
              }}
              onClick={ () => facebookLogin() }
              text={getString('SIGN_IN_WITH_FACEBOOK')}
            />

          </ButtonWrapper>
        </ContentScroll>
      </ContentWrapper>
    </Container>
  );
}

export default Intro;
