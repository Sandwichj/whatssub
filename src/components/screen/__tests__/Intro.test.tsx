import 'react-native';
import * as React from 'react';

// Note: test renderer must be required after react-native.
import { ThemeProvider } from 'styled-components/native';
import renderer from 'react-test-renderer';
import { render, fireEvent, waitForElement } from 'react-native-testing-library';

import { AppProvider } from '../../../providers';
import Intro from '../Intro';
import Button from '../../shared/Button';
import { createTheme, ThemeType } from '../../../theme';

const createTestProps = (obj: object) => ({
  navigation: {
    navigate: jest.fn(),
  },
  ...obj,
});

const props: any = createTestProps({});
const component = (
  <AppProvider>
    <ThemeProvider theme={createTheme(ThemeType.LIGHT)}>
      <Intro
        {...props}
        screenProps={{
          theme: createTheme(ThemeType.LIGHT),
        }}
      />
    </ThemeProvider>
  </AppProvider>
);
let json: renderer.ReactTestRendererJSON;

// test for the container page in dom
describe('[Intro] screen rendering test', () => {
  it('should render outer component and snapshot matches', () => {
    json = renderer.create(component).toJSON();
    expect(json).toMatchSnapshot();
  });
});

describe('[Intro] Interaction', () => {
  let rendered: renderer.ReactTestRenderer;
  let root: renderer.ReactTestInstance;
  let testingLib: any;

  it('should simulate [googleLogin] click', () => {
    rendered = renderer.create(component);
    root = rendered.root;
    testingLib = render(component);

    jest.useFakeTimers();
    const buttons = root.findAllByType(Button);
    fireEvent(testingLib.getByTestId('btnGoogle'), 'click');

    // expect(context.dispatch).toHaveBeenCalledWith({ type: 'reset-user' });
    // expect(context.dispatch).toHaveBeenCalledWith({ type: 'set-user' }, expect.any(Object));

    jest.runAllTimers();
    expect(buttons[0].props.isLoading).toEqual(false); // TODO: test with useState
  });

  it('should simulate [facebookLogin] click', () => {
    rendered = renderer.create(component);
    root = rendered.root;

    const buttons = root.findAllByType(Button);
    fireEvent(testingLib.getByTestId('btnFacebook'), 'click');
    expect(buttons[1].props.isLoading).toEqual(false);
  });
});
