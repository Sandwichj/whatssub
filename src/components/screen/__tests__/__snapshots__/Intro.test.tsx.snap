// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Intro] screen rendering test should render outer component and snapshot matches 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "alignSelf": "stretch",
        "backgroundColor": "#023059",
        "flexBasis": 0,
        "flexDirection": "column",
        "flexGrow": 1,
        "flexShrink": 1,
        "justifyContent": "flex-start",
        "overflow": "hidden",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
          "marginTop": 120,
        },
      ]
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../assets/icons/logo.png",
        }
      }
      style={
        Array [
          Object {
            "height": 30,
            "width": 135,
          },
        ]
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "flex-start",
          "flexDirection": "column",
          "height": "100%",
          "justifyContent": "center",
          "width": "100%",
        },
      ]
    }
  >
    <RCTScrollView
      contentContainerStyle={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        }
      }
      style={
        Array [
          Object {
            "height": "100%",
            "width": "100%",
          },
        ]
      }
    >
      <View>
        <Image
          source={
            Object {
              "testUri": "../../../assets/icons/slash.png",
            }
          }
          style={
            Array [
              Object {
                "height": 14,
                "width": 14,
              },
              Object {
                "marginBottom": 34,
                "marginTop": 68,
              },
            ]
          }
        />
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 36,
              "fontWeight": "600",
              "height": 40,
              "opacity": 0,
            }
          }
        >
          정기결제
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#ffffff",
                "fontSize": 36,
                "fontWeight": "600",
                "lineHeight": 48,
              },
            ]
          }
        >
          잊지말고
똑똑하게
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#ffffff",
                "fontSize": 36,
                "fontWeight": "600",
                "lineHeight": 48,
              },
              Object {
                "color": "#ff84a5",
              },
            ]
          }
        >
          왓-썹
        </Text>
        <Image
          source={
            Object {
              "testUri": "../../../assets/icons/slash.png",
            }
          }
          style={
            Array [
              Object {
                "height": 14,
                "width": 14,
              },
              Object {
                "marginBottom": 68,
                "marginTop": 38,
              },
            ]
          }
        />
      </View>
    </RCTScrollView>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "bottom": 40,
          "flexDirection": "column",
          "height": 200,
          "justifyContent": "flex-start",
          "paddingTop": 28,
          "position": "absolute",
          "width": "100%",
        },
      ]
    }
  >
    <View
      accessible={true}
      clickable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "alignSelf": "center",
              "backgroundColor": "#023059",
              "borderColor": "#023059",
              "borderRadius": 4,
              "borderWidth": 2,
              "height": 52,
              "justifyContent": "center",
              "width": 320,
            },
            Object {
              "backgroundColor": "white",
              "borderColor": "#e1e4e7",
              "borderRadius": 0,
              "borderWidth": 1,
            },
          ]
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../assets/icons/google.png",
            }
          }
          style={
            Array [
              Object {
                "height": 24,
                "left": 16,
                "position": "absolute",
                "width": 24,
              },
            ]
          }
        />
        <Text
          style={
            Array [
              Object {
                "color": "#323b43",
                "fontSize": 14,
              },
            ]
          }
        >
          Sign in with Google
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "marginTop": 8,
        }
      }
    />
    <View
      accessible={true}
      clickable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "alignSelf": "center",
              "backgroundColor": "#023059",
              "borderColor": "#023059",
              "borderRadius": 4,
              "borderWidth": 2,
              "height": 52,
              "justifyContent": "center",
              "width": 320,
            },
            Object {
              "backgroundColor": "white",
              "borderColor": "#e1e4e7",
              "borderRadius": 0,
              "borderWidth": 1,
            },
          ]
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../assets/icons/facebook.png",
            }
          }
          style={
            Array [
              Object {
                "height": 24,
                "left": 16,
                "position": "absolute",
                "width": 24,
              },
              Object {
                "height": 28,
                "width": 16,
              },
            ]
          }
        />
        <Text
          style={
            Array [
              Object {
                "color": "#323b43",
                "fontSize": 14,
              },
            ]
          }
        >
          Sign in with Facebook
        </Text>
      </View>
    </View>
  </View>
</View>
`;
